src/backend/cuda/kernel/homography.hpp

```diff
@@ -0,0 +1,700 @@
+    __syncthreads();
+}
+
+__device__ bool computeMeanScale(
+    float* x_src_mean,
+    float* y_src_mean,
+    float* x_dst_mean,
+    float* y_dst_mean,
+    float* src_scale,
+    float* dst_scale,
+    float* src_pt_x,
+    float* src_pt_y,
+    float* dst_pt_x,
+    float* dst_pt_y,
+    CParam<float> x_src,
+    CParam<float> y_src,
+    CParam<float> x_dst,
+    CParam<float> y_dst,
+    CParam<float> rnd,
+    int i)
+{
+    const unsigned ridx = rnd.dims[0] * i;
+    unsigned r[4] = { (unsigned)rnd.ptr[ridx],
+                      (unsigned)rnd.ptr[ridx+1],
+                      (unsigned)rnd.ptr[ridx+2],
+                      (unsigned)rnd.ptr[ridx+3] };
+
+    // If one of the points is repeated, it's a bad samples, will still
+    // compute homography to ensure all threads pass __syncthreads()
+    bool bad = (r[0] == r[1] || r[0] == r[2] || r[0] == r[3] ||
+                r[1] == r[2] || r[1] == r[3] || r[2] == r[3]);
+
+    for (unsigned j = 0; j < 4; j++) {
+        src_pt_x[j] = x_src.ptr[r[j]];
+        src_pt_y[j] = y_src.ptr[r[j]];
+        dst_pt_x[j] = x_dst.ptr[r[j]];
+        dst_pt_y[j] = y_dst.ptr[r[j]];
+    }
+
+    *x_src_mean = (src_pt_x[0] + src_pt_x[1] + src_pt_x[2] + src_pt_x[3]) / 4.f;
+    *y_src_mean = (src_pt_y[0] + src_pt_y[1] + src_pt_y[2] + src_pt_y[3]) / 4.f;
+    *x_dst_mean = (dst_pt_x[0] + dst_pt_x[1] + dst_pt_x[2] + dst_pt_x[3]) / 4.f;
+    *y_dst_mean = (dst_pt_y[0] + dst_pt_y[1] + dst_pt_y[2] + dst_pt_y[3]) / 4.f;
+
+    float src_var = 0.0f, dst_var = 0.0f;
+    for (unsigned j = 0; j < 4; j++) {
+        src_var += sq(src_pt_x[j] - *x_src_mean) + sq(src_pt_y[j] - *y_src_mean);
+        dst_var += sq(dst_pt_x[j] - *x_dst_mean) + sq(dst_pt_y[j] - *y_dst_mean);
+    }
+
+    src_var /= 4.f;
+    dst_var /= 4.f;
+
+    *src_scale = sqrt(2.0f) / sqrt(src_var);
+    *dst_scale = sqrt(2.0f) / sqrt(dst_var);
+
+    return !bad;
+}
+
+#define APTR(Z, Y, X) (A.ptr[(Z) * A.dims[0] * A.dims[1] + (Y) * A.dims[0] + (X)])
+
+template<typename T>
+__global__ void buildLinearSystem(
+    Param<T> H,
+    Param<T> A,
+    Param<T> V,
+    CParam<float> x_src,
+    CParam<float> y_src,
+    CParam<float> x_dst,
+    CParam<float> y_dst,
+    CParam<float> rnd,
+    const unsigned iterations)
+{
+    unsigned i = blockIdx.y * blockDim.y + threadIdx.y;
+
+    if (i < iterations) {
+        float x_src_mean, y_src_mean;
+        float x_dst_mean, y_dst_mean;
+        float src_scale, dst_scale;
+        float src_pt_x[4], src_pt_y[4], dst_pt_x[4], dst_pt_y[4];
+
+        computeMeanScale(&x_src_mean, &y_src_mean,
+                         &x_dst_mean, &y_dst_mean,
+                         &src_scale, &dst_scale,
+                         src_pt_x, src_pt_y,
+                         dst_pt_x, dst_pt_y,
+                         x_src, y_src, x_dst, y_dst,
+                         rnd, i);
+
+        // Compute input matrix
+        for (unsigned j = threadIdx.x; j < 4; j+=blockDim.x) {
```

Comment:
It won't execute more than once with the current group size, but I want to have this easily adjustable for future changes.
It makes no difference using int instead of unsigned.